src/nvim/os/lang.c

```diff
@@ -21,55 +22,24 @@ void lang_init(void)
 {
 #ifdef __APPLE__
   if (os_getenv("LANG") == NULL) {
-    const char *lang_region = NULL;
-    CFTypeRef cf_lang_region = NULL;
-
-    CFLocaleRef cf_locale = CFLocaleCopyCurrent();
-    if (cf_locale) {
-      cf_lang_region = CFLocaleGetValue(cf_locale, kCFLocaleIdentifier);
-      CFRetain(cf_lang_region);
-      lang_region = CFStringGetCStringPtr(cf_lang_region,
-                                          kCFStringEncodingUTF8);
-      CFRelease(cf_locale);
-    } else {
-      // Use the primary language defined in Preferences -> Language & Region
-      CFArrayRef cf_langs = CFLocaleCopyPreferredLanguages();
-      if (cf_langs && CFArrayGetCount(cf_langs) > 0) {
-        cf_lang_region = CFArrayGetValueAtIndex(cf_langs, 0);
-        CFRetain(cf_lang_region);
-        CFRelease(cf_langs);
-        lang_region = CFStringGetCStringPtr(cf_lang_region,
-                                            kCFStringEncodingUTF8);
-      } else {
-        ELOG("$LANG is empty and your primary language cannot be inferred.");
-        return;
-      }
-    }
-
     char buf[50] = { 0 };
-    bool set_lang;
-    if (lang_region) {
-      set_lang = true;
-      xstrlcpy(buf, lang_region, sizeof(buf));
-    } else {
-      set_lang = CFStringGetCString(cf_lang_region, buf, 40,
-                                    kCFStringEncodingUTF8);
-    }
-    if (set_lang) {
+
+    // $LANG is not set, either because it was unset or Nvim was started
+    // from the Dock. Query the system locale.
+    if (LocaleRefGetPartString(NULL,
+                               kLocaleLanguageMask | kLocaleLanguageVariantMask |
+                               kLocaleRegionMask | kLocaleRegionVariantMask,
+                               sizeof(buf) - 10, buf) == noErr && *buf) {
       if (strcasestr(buf, "utf-8") == NULL) {
         xstrlcat(buf, ".UTF-8", sizeof(buf));
       }
       os_setenv("LANG", buf, true);
+      setlocale(LC_ALL, "");
+      // Make sure strtod() uses a decimal point, not a comma.
+      setlocale(LC_NUMERIC, "C");
+    } else {
+      ELOG("$LANG is empty and the macOS primary language cannot be inferred.");
     }
-    CFRelease(cf_lang_region);
-# ifdef HAVE_LOCALE_H
```

Comment:
I deliberately left out the checks for `HAVE_LOCALE_H` and `LC_NUMERIC`, because I think it's not needed within `__APPLE__`.